File: crates/wh40kdc/src/import/newrecruit_json.rs
//! NewRecruit JSON adapter: lower a decoded NewRecruit roster export
//! (BattleScribe-derived tree, same outer shape as ListForge) to a
//! [`ParsedRoster`].
//!
//! NewRecruit-specific detection: `generatedBy` is `"https://newrecruit…"`
//! and/or `roster.xmlns` is the BattleScribe rosterSchema namespace.
//!
//! The primary faction surfaces in `forces[].catalogueName`
//! (e.g. "Chaos - Chaos Knights") — we take the segment after the final
//! `" - "`. Falls back to the first `"Faction: X"` category.
//!
//! The walk reads the same allowlist as the ListForge adapter — `name`,
//! `number`, `type`, `categories[].name`, `group`, `costs` point values, and
//! `catalogueName`. `rules`/`profiles`/`description` are never touched.
//!
//! Rust mirror of `tools/src/import/newrecruit-json.ts`.

use serde_json::Value;

use super::adapter::{FormatAdapter, ParseError};
use super::types::{
    AttachmentRole, ParsedLeaderAttachment, ParsedLoadoutGroup, ParsedRoster, ParsedUnit,
    ParsedWargear, RosterFormat,
};

const PTS_COST_NAME: &str = "pts";
const ENHANCEMENT_GROUP_PREFIX: &str = "Enhancements";
const ATTACHMENT_GROUP_PREFIX: &str = "40kdc Attachment:";
const CHARACTER_CATEGORIES: [&str; 2] = ["Character", "Epic Hero"];
const WEAPON_CATEGORY_SUFFIX: &str = " Weapon";
const NEWRECRUIT_XMLNS: &str = "http://www.battlescribe.net/schema/rosterSchema";
const NEWRECRUIT_HOST_PREFIX: &str = "https://newrecruit";

fn as_array(value: &Value) -> &[Value] {
    value.as_array().map(Vec::as_slice).unwrap_or(&[])
}

fn as_string(value: &Value) -> Option<&str> {
    value.as_str()
}

fn selection_name(sel: &Value) -> &str {
    as_string(&sel["name"]).unwrap_or("")
}

fn selection_type(sel: &Value) -> &str {
    as_string(&sel["type"]).unwrap_or("")
}

fn selection_count(sel: &Value) -> u64 {
    match sel["number"].as_u64() {
        Some(n) if n > 0 => n,
        _ => 1,
    }
}

fn points_of(sel: &Value) -> Option<u64> {
    for cost in as_array(&sel["costs"]) {
        if as_string(&cost["name"]) == Some(PTS_COST_NAME) {
            if let Some(v) = cost["value"].as_f64() {
                return Some(v as u64);
            }
        }
    }
    None
}

fn category_names(sel: &Value) -> Vec<&str> {
    as_array(&sel["categories"])
        .iter()
        .filter_map(|c| as_string(&c["name"]))
        .collect()
}

fn child_selections(sel: &Value) -> &[Value] {
    as_array(&sel["selections"])
}

fn walk(sel: &Value, visit: &mut impl FnMut(&Value)) {
    visit(sel);
    for child in child_selections(sel) {
        walk(child, visit);
    }
}

fn is_unit_selection(sel: &Value) -> bool {
    matches!(selection_type(sel), "model" | "unit")
}

fn is_character(sel: &Value) -> bool {
    category_names(sel)
        .iter()
        .any(|n| CHARACTER_CATEGORIES.contains(n))
}

fn keyword_overrides(sel: &Value) -> Vec<String> {
    as_array(&sel["categories"])
        .iter()
        .filter(|category| {
            as_string(&category["id"]).is_some_and(|id| id.starts_with("40kdc-keyword-"))
        })
        .filter_map(|category| as_string(&category["name"]))
        .map(str::to_string)
        .collect()
}

fn attachment_of(sel: &Value) -> Option<ParsedLeaderAttachment> {
    let group = as_string(&sel["group"])?;
    let suffix = group.strip_prefix(ATTACHMENT_GROUP_PREFIX)?;
    let mut parts = suffix.split(':');
    let role = match parts.next()? {
        "leader" => AttachmentRole::Leader,
        "support" => AttachmentRole::Support,
        _ => return None,
    };
    let provisional = match parts.next()? {
        "confirmed" => false,
        "provisional" => true,
        _ => return None,
    };
    if parts.next().is_some() {
        return None;
    }
    Some(ParsedLeaderAttachment {
        bodyguard_raw_name: selection_name(sel).to_string(),
        role,
        provisional,
    })
}

fn is_weapon_selection(sel: &Value) -> bool {
    category_names(sel)
        .iter()
        .any(|n| n.ends_with(WEAPON_CATEGORY_SUFFIX))
}

fn is_enhancement_selection(sel: &Value) -> bool {
    as_string(&sel["group"]).is_some_and(|g| g.starts_with(ENHANCEMENT_GROUP_PREFIX))
}

fn model_count(unit: &Value) -> u64 {
    let mut total = 0;
    walk(unit, &mut |s| {
        if selection_type(s) == "model" {
            total += selection_count(s);
        }
    });
    if total > 0 {
        total
    } else {
        selection_count(unit)
    }
}

fn parse_unit(unit: &Value) -> ParsedUnit {
    let mut wargear: Vec<ParsedWargear> = Vec::new();
    let mut enhancement_raw_name: Option<String> = None;
    let mut enhancement_points: Option<u64> = None;
    let mut leader_attachment: Option<ParsedLeaderAttachment> = None;
    let mut is_warlord = false;

    for node in child_selections(unit) {
        walk(node, &mut |s| {
            if let Some(attachment) = attachment_of(s) {
                leader_attachment = Some(attachment);
                return;
            }
            if is_enhancement_selection(s) {
                if enhancement_raw_name.is_none() {
                    enhancement_raw_name = Some(selection_name(s).to_string());
                    enhancement_points = points_of(s);
                }
                return;
            }
            if selection_name(s) == "Warlord" {
                is_warlord = true;
                return;
            }
            if is_weapon_selection(s) {
                wargear.push(ParsedWargear {
                    raw_name: selection_name(s).to_string(),
                    count: selection_count(s),
                });
            }
        });
    }

    // Model selections retain an exact per-model breakdown. Only preserve a
    // group when every nested weapon total divides across its models; otherwise
    // resolving the aggregate is the only faithful representation.
    let loadout_groups: Vec<ParsedLoadoutGroup> = child_selections(unit)
        .iter()
        .filter(|model| selection_type(model) == "model")
        .filter_map(|model| {
            let count = selection_count(model);
            let mut totals = Vec::new();
            walk(model, &mut |selection| {
                if is_weapon_selection(selection) {
                    totals.push(ParsedWargear {
                        raw_name: selection_name(selection).to_string(),
                        count: selection_count(selection),
                    });
                }
            });
            if count == 0 || totals.iter().any(|item| item.count % count != 0) {
                return None;
            }
            Some(ParsedLoadoutGroup {
                model_name: Some(selection_name(model).to_string()),
                count,
                wargear: totals
                    .into_iter()
                    .map(|item| ParsedWargear {
                        raw_name: item.raw_name,
                        count: item.count / count,
                    })
                    .collect(),
            })
        })
        .collect();

    let loadout_groups = (!loadout_groups.is_empty()).then_some(loadout_groups);

    ParsedUnit {
        raw_name: selection_name(unit).to_string(),
        is_character: is_character(unit),
        keyword_overrides: Some(keyword_overrides(unit)),
        model_count: model_count(unit),
        points: points_of(unit),
        is_warlord,
        enhancement_raw_name,
        enhancement_points,
        wargear,
        loadout_groups,
        leader_attachment: Some(leader_attachment),
    }
}
fn config_value(selections: &[Value], config_name: &str) -> Option<String> {
    let node = selections
        .iter()
        .find(|s| selection_name(s) == config_name)?;
    let child = child_selections(node).first()?;
    Some(selection_name(child).to_string())
}

/// Every value under a named config, across repeated blocks and multiple
/// children, in source order. Used for multi-detachment 11e lists.
fn config_values(selections: &[Value], config_name: &str) -> Vec<String> {
    let mut out = Vec::new();
    for node in selections {
        if selection_name(node) != config_name {
            continue;
        }
        for child in child_selections(node) {
            let name = selection_name(child);
            if !name.is_empty() {
                out.push(name.to_string());
            }
        }
    }
    out
}

fn parse_limit(label: Option<&str>) -> Option<u64> {
    let label = label?;
    let bytes = label.as_bytes();
    let lower = label.to_ascii_lowercase();
    let mut i = 0;
    while i < bytes.len() {
        if bytes[i].is_ascii_digit() {
            let start = i;
            while i < bytes.len() && (bytes[i].is_ascii_digit() || bytes[i] == b',') {
                i += 1;
            }
            let mut j = i;
            while j < bytes.len() && bytes[j] == b' ' {
                j += 1;
            }
            if lower[j..].starts_with("point") {
                let digits: String = label[start..i].chars().filter(|c| *c != ',').collect();
                return digits.parse().ok();
            }
        } else {
            i += 1;
        }
    }
    None
}

fn collect_factions(forces: &[Value]) -> Vec<String> {
    let mut seen: Vec<String> = Vec::new();
    for force in forces {
        for sel in child_selections(force) {
            walk(sel, &mut |s| {
                for name in category_names(s) {
                    if let Some(rest) = name.strip_prefix("Faction:") {
                        let value = rest.trim().to_string();
                        if !value.is_empty() && !seen.contains(&value) {
                            seen.push(value);
                        }
                    }
                }
            });
        }
    }
    seen
}

/// Primary faction from a force's `catalogueName` (e.g.
/// "Chaos - Chaos Knights" → "Chaos Knights").
fn primary_faction_from_catalogue(forces: &[Value]) -> Option<String> {
    for force in forces {
        let name = as_string(&force["catalogueName"])?;
        let last = name.rsplit(" - ").next()?.trim();
        if !last.is_empty() {
            return Some(last.to_string());
        }
    }
    None
}

fn roster_of(decoded: &Value) -> Option<&Value> {
    let roster = decoded.get("roster")?;
    if !roster.is_object() {
        return None;
    }
    if !roster.get("forces").map(Value::is_array).unwrap_or(false) {
        return None;
    }
    Some(roster)
}

fn has_newrecruit_signature(decoded: &Value, roster: &Value) -> bool {
    if as_string(&roster["xmlns"]) == Some(NEWRECRUIT_XMLNS) {
        return true;
    }
    let gen_by = as_string(&decoded["generatedBy"]).or_else(|| as_string(&roster["generatedBy"]));
    if let Some(g) = gen_by {
        if g.to_ascii_lowercase().starts_with(NEWRECRUIT_HOST_PREFIX) {
            return true;
        }
    }
    false
}

pub struct NewRecruitJsonAdapter;

impl FormatAdapter for NewRecruitJsonAdapter {
    fn format(&self) -> RosterFormat {
        RosterFormat::NewrecruitJson
    }

    fn detect(&self, decoded: &Value) -> bool {
        match roster_of(decoded) {
            Some(roster) => has_newrecruit_signature(decoded, roster),
            None => false,
        }
    }

    fn parse(&self, decoded: &Value) -> Result<ParsedRoster, ParseError> {
        let roster = roster_of(decoded).ok_or_else(|| {
            ParseError("newrecruit-json: payload has no roster.forces array".into())
        })?;
        let forces = as_array(&roster["forces"]);

        let mut detachment_raw_names: Vec<String> = Vec::new();
        let mut battle_size_raw: Option<String> = None;
        let mut force_disposition_raw_name: Option<String> = None;
        let mut units: Vec<ParsedUnit> = Vec::new();
        for force in forces {
            let top = child_selections(force);
            detachment_raw_names.extend(config_values(top, "Detachment"));
            if battle_size_raw.is_none() {
                battle_size_raw = config_value(top, "Battle Size");
            }
            if force_disposition_raw_name.is_none() {
                force_disposition_raw_name = config_value(top, "Force Disposition");
            }
            for sel in top {
                if is_unit_selection(sel) {
                    units.push(parse_unit(sel));
                }
            }
        }

        let factions = collect_factions(forces);
        let primary_faction =
            primary_faction_from_catalogue(forces).or_else(|| factions.first().cloned());
        let total_reported = points_of(roster);

        let mut total_computed = 0;
        for force in forces {
            for sel in child_selections(force) {
                walk(sel, &mut |s| {
                    if let Some(pts) = points_of(s) {
                        total_computed += pts;
                    }
                });
            }
        }

        let generated_by = as_string(&decoded["generatedBy"])
            .or_else(|| as_string(&roster["generatedBy"]))
            .map(str::to_string);

        let name = as_string(&decoded["name"])
            .or_else(|| as_string(&roster["name"]))
            .unwrap_or("Imported roster")
            .to_string();

        Ok(ParsedRoster {
            name,
            generated_by,
            faction_raw_name: primary_faction,
            detachment_raw_names,
            battle_size_raw: battle_size_raw.clone(),
            force_disposition: None,
            force_disposition_raw_name: Some(force_disposition_raw_name),
            declared_limit: parse_limit(battle_size_raw.as_deref()),
            total_reported,
            total_computed,
            units,
            multi_force: factions.len() > 1,
        })
    }
}

File: python/src/wh40kdc/export/newrecruit_json.py
"""NewRecruit JSON exporter.

Emits a BattleScribe-shaped roster skeleton that round-trips through the
NewRecruit JSON adapter. The shape carries only fields the importer reads:
``name``, ``type``, ``number``, ``costs[]``, ``categories[].name``, ``group``,
and ``catalogueName``. No ``rules`` / ``profiles`` / ``description`` ever
appear — we don't store them, and emitting them would be an IP violation.

Faction and detachment display names come from ``title_case_id(faction_id)``
— the Roster doesn't carry the source's raw faction name, so we reconstruct
it from the kebab-case id. This is the only lossy hop in the JSON round-trip
(e.g. ``tau-empire`` → "Tau Empire" rather than the canonical "T'au Empire").

Python mirror of ``tools/src/export/newrecruit-json.ts``; byte-identical
output, so dict insertion order replicates the TS object field order exactly
and conditional fields are omitted (never ``null``).
"""

from __future__ import annotations

from typing import Any

from wh40kdc.export.helpers import (
    Roster,
    RosterUnit,
    pretty_json,
    title_case_id,
    total_army_points,
)

_PTS_TYPE_ID = "pts-type"
_NEWRECRUIT_XMLNS = "http://www.battlescribe.net/schema/rosterSchema"
_NEWRECRUIT_GENERATED_BY = "https://newrecruit.eu"


def _faction_category(roster: Roster) -> dict[str, Any] | None:
    """Build a "Faction: <name>" category from the unit's roster context."""
    display = title_case_id(roster.get("faction_id"))
    if display is None:
        return None
    return {"name": f"Faction: {display}", "primary": False}


def _wargear_selection(idx: int, w: dict[str, Any]) -> dict[str, Any]:
    return {
        "id": f"w-{idx}",
        "name": w["ref"]["raw_name"],
        "type": "upgrade",
        "number": w["count"],
        # The NewRecruit importer recognises a wargear selection by a category
        # ending in " Weapon" — emit a generic "Ranged Weapon" so we don't
        # have to track ranged-vs-melee separation the Roster doesn't model.
        "categories": [{"name": "Ranged Weapon", "primary": False}],
    }


def _group_model_selection(
    unit_index: int, group_index: int, unit: RosterUnit, group: dict[str, Any]
) -> dict[str, Any]:
    model_name = group.get("model_name")
    if model_name is None:
        model_name = unit["ref"]["raw_name"]
    return {
        "id": f"u{unit_index}-model-{group_index}",
        "name": model_name,
        "type": "model",
        "number": group["count"],
        "selections": [
            _wargear_selection(
                item_index,
                {**item, "count": item["count"] * group["count"]},
            )
            for item_index, item in enumerate(group["wargear"])
        ],
    }


def _unit_selection(idx: int, u: RosterUnit, faction: dict[str, Any] | None) -> dict[str, Any]:
    inner: list[dict[str, Any]] = []
    if u.get("is_warlord"):
        inner.append({"id": f"u{idx}-warlord", "name": "Warlord", "type": "upgrade", "number": 1})
    if u.get("leader_attachment"):
        attachment = u["leader_attachment"]
        status = "provisional" if attachment["provisional"] else "confirmed"
        inner.append(
            {
                "id": f"u{idx}-attachment",
                "name": attachment["bodyguard_ref"]["raw_name"],
                "type": "upgrade",
                "number": 1,
                "group": f"40kdc Attachment:{attachment['role']}:{status}",
            }
        )
    if u.get("enhancement"):
        enh: dict[str, Any] = {
            "id": f"u{idx}-enh",
            "name": u["enhancement"]["raw_name"],
            "type": "upgrade",
            "number": 1,
            "group": "Enhancements",
        }
        if u.get("enhancement_points") is not None:
            enh["costs"] = [
                {"name": "pts", "typeId": _PTS_TYPE_ID, "value": u["enhancement_points"]}
            ]
        inner.append(enh)

    wargear_selections = [_wargear_selection(wi, w) for wi, w in enumerate(u["wargear"])]
    groups = u.get("loadout_groups") or []
    if groups:
        model_selections = [
            _group_model_selection(idx, group_index, u, group)
            for group_index, group in enumerate(groups)
        ]
    else:
        model_selections = [
            {
                "id": f"u{idx}-model",
                "name": u["ref"]["raw_name"],
                "type": "model",
                "number": u["model_count"],
                "selections": wargear_selections,
            }
        ]
    own_categories = [
        *([faction] if faction else []),
        *[
            {"id": f"40kdc-keyword-{index}", "name": keyword, "primary": False}
            for index, keyword in enumerate(u.get("keyword_overrides") or [])
        ],
    ]
    if u["model_count"] <= 1 and not groups:
        sel: dict[str, Any] = {
            "id": f"u-{idx}",
            "name": u["ref"]["raw_name"],
            "type": "model",
            "number": 1,
            "categories": own_categories,
        }
        if u["points"] is not None:
            sel["costs"] = [{"name": "pts", "typeId": _PTS_TYPE_ID, "value": u["points"]}]
        sel["selections"] = [*inner, *wargear_selections]
        return sel
    sel = {
        "id": f"u-{idx}",
        "name": u["ref"]["raw_name"],
        "type": "unit",
        "number": 1,
        "categories": own_categories,
    }
    if u["points"] is not None:
        sel["costs"] = [{"name": "pts", "typeId": _PTS_TYPE_ID, "value": u["points"]}]
    sel["selections"] = [*inner, *model_selections]
    return sel


def _config_selection(name: str, value: str, idx: str) -> dict[str, Any]:
    return {
        "id": f"cfg-{idx}",
        "name": name,
        "type": "upgrade",
        "number": 1,
        "categories": [{"name": "Configuration", "primary": True}],
        "selections": [
            {
                "id": f"cfg-{idx}-val",
                "name": value,
                "type": "upgrade",
                "number": 1,
            }
        ],
    }


def _battle_size_label(roster: Roster) -> str | None:
    declared_limit = roster["points"].get("declared_limit")
    if roster.get("battle_size") == "strike-force":
        limit = declared_limit if declared_limit is not None else 2000
        return f"Strike Force ({limit} Point limit)"
    if roster.get("battle_size") == "incursion":
        limit = declared_limit if declared_limit is not None else 1000
        return f"Incursion ({limit} Point limit)"
    return None


def serialize_newrecruit_json(roster: Roster) -> str:
    faction = _faction_category(roster)
    faction_display = title_case_id(roster.get("faction_id"))
    if faction_display is None:
        faction_display = "Unknown"
    battle_size = _battle_size_label(roster)

    config: list[dict[str, Any]] = []
    if battle_size:
        config.append(_config_selection("Battle Size", battle_size, "battle-size"))
    for d in roster["detachments"]:
        config.append(_config_selection("Detachment", d["ref"]["raw_name"], "detachment"))
    force_disposition = roster.get("force_disposition")
    if force_disposition is not None:
        config.append(
            _config_selection(
                "Force Disposition",
                title_case_id(force_disposition) or force_disposition,
                "force-disposition",
            )
        )

    force = {
        "id": "force-1",
        "name": "Army Roster",
        "catalogueName": faction_display,
        "selections": [
            *config,
            *(_unit_selection(i, u, faction) for i, u in enumerate(roster["units"])),
        ],
    }

    total = total_army_points(roster)

    payload = {
        "name": roster["name"],
        "generatedBy": _NEWRECRUIT_GENERATED_BY,
        "roster": {
            "name": roster["name"],
            "xmlns": _NEWRECRUIT_XMLNS,
            "generatedBy": _NEWRECRUIT_GENERATED_BY,
            "costs": [{"name": "pts", "typeId": _PTS_TYPE_ID, "value": total}],
            "forces": [force],
        },
    }

    return pretty_json(payload)
